dask_ml/cluster/spectral.py

```diff
@@ -208,93 +237,121 @@ def fit(self, X, y=None):
         # recover the original order.
         inds_idx = np.argsort(inds)
 
-        params = self.kernel_params or {}
-        params['gamma'] = self.gamma
-        params['degree'] = self.degree
-        params['coef0'] = self.coef0
-
         # compute the exact blocks
         # these are done in parallel for dask arrays
         if isinstance(X, da.Array):
-            X_keep = X[keep].rechunk(self.n_components).persist()
+            X_keep = X[keep].rechunk(X.shape).persist()
         else:
             X_keep = X[keep]
 
-        if isinstance(metric, six.string_types):
-            if metric not in PAIRWISE_KERNEL_FUNCTIONS:
-                msg = ("Unknown affinity metric name '{}'. Expected one "
-                       "of '{}'".format(metric,
-                                        PAIRWISE_KERNEL_FUNCTIONS.keys()))
-                raise ValueError(msg)
-            A = pairwise_kernels(X_keep,
-                                 metric=metric, filter_params=True, **params)
-            B = pairwise_kernels(X_keep, X[rest],
-                                 metric=metric, filter_params=True, **params)
-
-        elif callable(metric):
-            A = metric(X_keep, **params)
-            B = metric(X_keep, X[rest], **params)
-        else:
-            msg = ("Unexpected type for 'affinity' '{}'. Must be string "
-                   "kernel name, array, or callable")
-            raise TypeError(msg)
-        if isinstance(A, da.Array):
-            A = A.rechunk((n_components, n_components))
-            B = B.rechunk((B.shape[0], B.chunks[1]))
+        X_rest = X[rest]
+
+        chunks = rechunk_strategy.get('kernel')
+        if chunks:
+            logger.info("Rechunking for kernel %s -> %s",
+                        X_rest.numblocks, chunks)
+            X_rest = X_rest.rechunk(chunks)
+
+        A, B = embed(X_keep, X_rest, n_components, metric, params)
+        _log_array(logger, A, 'A')
+        _log_array(logger, B, 'B')
 
         # now the approximation of C
         a = A.sum(0)   # (l,)
         b1 = B.sum(1)  # (l,)
         b2 = B.sum(0)  # (m,)
 
+        # TODO: I think we have some unnecessary delayed wrapping of A here.
         A_inv = da.from_delayed(delayed(pinv)(A), A.shape, A.dtype)
```

Comment:
A is a dask array here or a numpy array?

If it's a dask array then what is your objective here?  Do you want to collapse it into a single chunk and then apply `pinv`?

Reply:
> A is a dask array here or a numpy array?

Could be either depending on whether the original `X` is a dask or numpy array (I think).